bionic/code_hasher.py

```diff
@@ -66,7 +69,7 @@ def _ingest_raw_prefix_and_bytes(self, type_prefix, obj_bytes=b""):
         self._hash.update(PREFIX_SEPARATOR)
         self._hash.update(obj_bytes)
 
-    def _check_and_ingest(self, obj):
+    def _check_and_ingest(self, obj, code_context=None):
```

Comment:
`code_context` will always be `None` for the first call. Do you think it's easier to read if we pass a `DEFAULT_CONTEXT` instead?

Reply:
Oh, you're right, just requiring it to be non-`None` won't work. Let me rephrase my worry:

`code_context` is sometimes `None` and sometimes not, and there are some code paths that assume that it's not `None`. So how are we guaranteeing that it will always be present at the time that it's needed? I'm sure there's a reason it works, but it's not immediately apparent; I'd have to trace through all the variations to be convinced. Ideally the code would be structured so that it's obviously impossible for it to be missing when we need it, but we could also just have a comment that explains why this is safe.

(BTW, if we call `CodeHasher.hash` on a code object, will it work?)

Reply:
Oh I see. We really just need `code_context` to be present for code objects. We will get code objects only from a coroutine in the current implementation, and that will always create a code object for you. I can add an assert in `_ingest_code` method that `code_context` is not null, since that is the only place where we actually use the `code_context`. Wdyt?

To answer your question, `CodeHasher.hash` won't work on a code object. That would throw an error.

Reply:
Is there anything sensible we could do if we got passed a code object? Or is throwing an exception our only option?

If the former, then a `DEFAULT_CONTEXT` might make sense. If not, then I agree putting an assert there makes sense. (Plus a comment explaining why it can't happen, of course.)

Reply:
We can hash only the code object and skip the references. I don't think we can find references in that case. I'll add a `DEFAULT_CONTEXT` and add a field called `can_find_references` in `CodeContext`.